src/engine/renderer/gl_shader.cpp

```diff
@@ -1248,21 +1244,24 @@ void GLShaderManager::CompileGPUShaders( GLShader *shader, shaderProgram_t *prog
 	}
 }
 
-void GLShaderManager::CompileAndLinkGPUShaderProgram( GLShader *shader, shaderProgram_t *program,
+void GLShaderManager::CompileAndLinkGPUShaderProgram( GLShader *shader, shaderProgram_t *program, shaderProgram_t* baseProgram,
 						      Str::StringRef compileMacros, int deformIndex )
 {
-	GLShaderManager::CompileGPUShaders( shader, program, compileMacros );
+	if ( ( !baseProgram->VS && shader->_hasVertexShader ) || ( !baseProgram->FS && shader->_hasFragmentShader )
```

Comment:
It's kinda confusing that program and baseProgram can be the same. I'm not convinced that this refactoring is really simpler.

Reply:
Alternatively, code from `CompileAndLinkGPUShaderProgram()` can be moved to `buildPermutation()` and `baseProgram` set to `program` explicitly for `deformIndex == 0`.